File: platforms/neoforge-1.21.1/src/main/java/heckerpowered/lethal/platform/interop/MixinInterop.java
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 heckerpowered
 */

package heckerpowered.lethal.platform.interop;

import org.jetbrains.annotations.NotNull;

public final class MixinInterop {
    private MixinInterop() {
    }

    @NotNull
    public static <Access> Access requireAccess(@NotNull Object host, @NotNull Class<Access> accessType) {
        if (!accessType.isInstance(host)) {
            throw new IllegalStateException(
                    "Required Mixin did not apply: " + host.getClass().getName() +
                            " does not implement " + accessType.getName()
            );
        }

        return accessType.cast(host);
    }

    @NotNull
    public static <Host> Host requireHost(@NotNull Object access, @NotNull Class<Host> hostType) {
        if (!hostType.isInstance(access)) {
            throw new IllegalArgumentException(
                    "Bridge access is not backed by the required NeoForge host type: " +
                            access.getClass().getName() + " is not " + hostType.getName()
            );
        }

        return hostType.cast(access);
    }
}

File: bridge/src/main/kotlin/heckerpowered/bridge/MixinInterop.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 heckerpowered
 */

package heckerpowered.bridge

/**
 * Verifies the shared JVM identity used by host objects enhanced through Mixin.
 *
 * These conversions do not create adapters. The host object and its bridge view must be the same object, so a failed
 * cast reports a broken platform implementation at the boundary where it is first observed.
 */
object MixinInterop {
    @JvmStatic
    fun <Access : Any> requireAccess(host: Any, accessType: Class<Access>): Access = accessType.cast(host)

    @JvmStatic
    fun <Host : Any> requireHost(access: Any, hostType: Class<Host>): Host = hostType.cast(access)
}

inline fun <reified Access : Any> requireAccess(host: Any): Access = MixinInterop.requireAccess(host, Access::class.java)

inline fun <reified Host : Any> requireHost(access: Any): Host = MixinInterop.requireHost(access, Host::class.java)
